src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -303,6 +413,34 @@ private string CreateTokenPrivate(JObject payload, SigningCredentials signingCre
 
                 rawHeader = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None)));
                 JsonWebTokenManager.KeyToHeaderCache.TryAdd(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), rawHeader);
+            } // Otherwise, if there is no outer JWT header, add additional header claims to this header.
+            else if (additionalHeaderClaims != null)
+            {
+                JObject header;
+                if (signingCredentials != null)
+                {
+                    header = new JObject
```

Comment:
nit: The process of creating a default header is repeated twice in this method. Perhaps we should add an internal utility method that creates a default header using singing credentials.

Reply:
Good suggestion! :) I added utility methods for creating default JWE and JWS headers.